arch/x86: Unbreak SMP startup on x86_64

andyross opened the pull request:
A last minute "cleanup" to the EFI startup path (on a system where I
had SMP disabled) moved the load of the x86_cpuboot[0] entry into RBP
into the main startup code, which is wrong because on auxiliary CPUs
that's already set up by the 16/32 bit entry code to point to the
OTHER entries.

Put it back where it belongs.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
arch/x86: Unbreak SMP startup on x86_64

A last minute "cleanup" to the EFI startup path (on a system where I
had SMP disabled) moved the load of the x86_cpuboot[0] entry into RBP
into the main startup code, which is wrong because on auxiliary CPUs
that's already set up by the 16/32 bit entry code to point to the
OTHER entries.

Put it back where it belongs.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented:
hmm one test fails, doesn't look familiar (to me anyway):
tests/kernel/smp/kernel.multiprocessing.smp in qemu_x86_64:tests/kernel/smp/kernel.multiprocessing
```
START - test_smp_ipi
cpu num=2
    Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/smp/src/main.c:532: test_smp_ipi: (sched_ipi_has_called == 0 is false)
shouldn't receive IPI,(1)
 FAIL - test_smp_ipi
```

andyross commented:
Hrm, another weird one.  This is working on my box, both the version pushed here and one rebased over master just now.  That is indeed a brand new test, but looking at the code it's REALLY simple and it's hard to imagine an IPI just plain failing like that...

andrewboie commented:
This must be a time slippage thing? local runs of just the test passed 6 out of 7 times. it only failed once. the initial failure I saw was a sanitycheck that had my system loaded up.

andyross commented:
Yeah, it looks like host timing yet again.  The test does a busy wait of 50ms after sending the IPI for the other core to receive it, but I'm guessing delivering that requires waking up a suspended thread, and on loaded systems with CPU bound threads (and this test is making it incrementally worse by busy waiting instead of going to idle on the initial core) that's probably long enough to trip over.   I can't make it work on my big system, but I did see it once in a sanitycheck run on my laptop.  I'm trying the patch below right now and guessing it resolves it.  If you have bandwidth to try it that would help too.  I'll add the fix to this PR once I think it works.

```diff
diff --git a/tests/kernel/smp/src/main.c b/tests/kernel/smp/src/main.c
index c3c4171d6e..b2d129bfa8 100644
--- a/tests/kernel/smp/src/main.c
+++ b/tests/kernel/smp/src/main.c
@@ -526,8 +526,6 @@ void test_smp_ipi(void)
 
        sched_ipi_has_called = 0;
 
-       k_busy_wait(DELAY_US);
-
        /* It shouldn't enter our IPI interrupt handler at this moment */
        zassert_true(sched_ipi_has_called == 0, "shouldn't receive IPI,(%d)",
                        sched_ipi_has_called);
@@ -536,8 +534,11 @@ void test_smp_ipi(void)
                /* issue a sched ipi to tell other CPU to run thread */
                arch_sched_ipi();
 
-               /* do busy wait here until get IPI */
-               k_busy_wait(DELAY_US);
+               /* Need to wait longer than we think, loaded CI
+                * systems need to wait for host scheduling to run the
+                * other CPU's thread.
+                */
+               k_msleep(300);
 
                /**TESTPOINT: check if enter our IPI interrupt handler */
                zassert_true(sched_ipi_has_called == i,
```

andyross commented:
Ah, heh.  It's more fundamental actually.  The test was written to check for a specific IPI being delivered, but it's implemented with a hook of all IPIs.  Arbitrary host scheduling also generates them, and there's no way to lock out the other CPU from doing any scheduling (the platform may have threads, the logging subsystem has a thread, and we obviously can't use 1cpu here because it's a test of MP). 

I should have thought harder during review, really this test case needed to have been added to the tests/kernel/mp test (which is NOT SMP and doesn't run Zephyr threads on the other CPU it's testing) and not here.

Nonetheless, in context this is coverage and we just want to know whether at least one IPI was delivered, if the scheduler IPI also fired that's not an error.  Simple patch inbound once I let it run for a few iterations.

Commit:
tests/kernel/smp: Fixup IPI test

This test was written to assume that the only IPI handled would be the
one generated by the test, but the scheduler also generates an IPI any
time a thread becomes runnable, and there's no way to lock that out in
an SMP system where the other CPU is going to be doing its own thing
(we can't use "1cpu" because that locks interrupts on the other CPU
and obviously this is a test of an interrupt).

Change the logic to detect that "at least one IPI was received", which
is fine for coverage.  Really a better place for a test like this
would have been tests/kernel/mp, which is a test of the lower level
APIs and runs the other CPU deterministically (i.e. not under the
control of the Zephyr scheduler).

Also some misc fixes:

* Don't busy wait at the start, that's needless.

* Sleep instead of busywaiting after sending the IPI, spinning isn't
  needed here and acts to increase CI load needlessly.

* Declare the cross thread signal variable volatile for correctness
  (though this error seems to have been benign in practice).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
OK, test fix added.  Seems to work for me.  Please re-review to cover the new patch.
